packages/aws-cdk-lib/core/lib/tag-aspect.ts

```diff
@@ -155,12 +158,41 @@ export class Tags {
     return new Tags(scope);
   }
 
-  private constructor(private readonly scope: IConstruct) { }
+  private readonly explicitStackTags: boolean;
+
+  private constructor(private readonly scope: IConstruct) {
+    this.explicitStackTags = FeatureFlags.of(scope).isEnabled(cxapi.EXPLICIT_STACK_TAGS) ?? false;
+  }
 
   /**
-   * add tags to the node of a construct and all its the taggable children
+   * Add tags to the node of a construct and all its the taggable children
+   *
+   * ## Tagging and CloudFormation Stacks
+   *
+   * If the feature flag `@aws-cdk/core:explicitStackTags` is set to `true`
+   * (recommended modern behavior), Stacks will not automatically be tagged.
+   * Stack tags should be configured on Stacks directly (preferred), or
+   * you must explicitly include the resource type `aws:cdk:stack` in the
+   * `includeResourceTypes` array.
+   *
+   * If the feature flag is set to `false` (legacy behavior) then both Stacks
+   * and resources in the indicated scope will both be tagged by default, which
+   * leads to tags being applied twice (once in the template, and then once
+   * again automatically by CloudFormation as part of the stack deployment).
+   * That behavior leads to loss of control as `excludeResourceTypes` will
+   * prevent tags from appearing in the template, but they will still be
+   * applied to the Stack and hence CloudFormation will still apply them
+   * to the resource.
    */
   public add(key: string, value: string, props: TagProps = {}) {
+    // Implicitly add `aws:cdk:stack` to the `excludeResourceTypes` array in modern behavior
+    if (this.explicitStackTags && !props.includeResourceTypes?.includes('aws:cdk:stack')) {
```

Comment:
If `props.includeResourceTypes?.includes('aws:cdk:stack')` is true, is there still a point in having the aspect traverse the tree? Can't we call `addStackTag()` to the stack and stop?